docker/docker_client.go

```diff
@@ -533,7 +533,8 @@ func (c *dockerClient) getBearerToken(ctx context.Context, challenge challenge,
 	defer res.Body.Close()
 	switch res.StatusCode {
 	case http.StatusUnauthorized:
-		return nil, ErrUnauthorizedForCredentials
+		err := client.HandleErrorResponse(res)
+		return nil, err
```

Comment:
This is not implausible, but both Buildah and Podman handle `ErrUnauthorizedForCredentials` specially. What about that?

Either `ErrUnauthorizedForCredentials` should be replaced/extended with some type that can preserve the server-provided status (and then all users need to be updated, for an API break), _and_ the other user of `ErrUnauthorizedForCredentials` must be updated a well — which is the long-term right thing to do;

Or do the small non-fix discussed in https://github.com/containers/libpod/issues/3884 , decode the server output, log it to `logrus`, but keep returning `ErrUnauthorizedForCredentials` to callers.

Reply:
how to update the usage of ErrUnauthorizedForCredentials? can I replace the  ErrUnauthorizedForCredentials with actual server side error and remove the check  `case docker.ErrUnauthorizedForCredentials:` from Podman and Buildah?

Reply:
To an extent, that depends on what Podman/Buildah want to do.  Right now, both use the `ErrUnauthorizedForCredentials` value to provide a “nice” error message, but throw away the underlying cause entirely.

It’s plausible that they could just report the original server-side value while removing that extra check, but that also depends on that value being nice enough for _most_ registries. 
Alternatively, we could replace the `ErrUnauthorizedForCredentials` error _value_ with an error _type_ that can still carry the underlying error returned by `client.HandleErrorResponse`. Then Podman/Buildah would still need updating, but they could still choose to use their own error texts.

I suppose we should collect sample responses from the major registries, and decide whether they are good enough to present without extra context.